### ISSUE TITLE
OOM on headers downloading

### ISSUE THREAD

**Issue by AskAlexSharov** [ledgerwatch/erigon#8212]
this pic is from mumbai, but i saw some reports from other nets in discord

![mem](https://github.com/ledgerwatch/erigon/assets/46885206/8b1dd93c-b9a7-4412-bc5b-bc799a97b91a)

**Comment by schur** [comment]
I cannot help with any detailed code analysis, but I can confirm the following
(based on 2 running instances of erigon, ethereum mainnet on separate servers each with their own config file and data directory):

- v2.48.1 and below have been running on my servers fine for months, no memory problems
- v2.49 and higher eat up all available system memory within around 1 day, then the process gets killed by the OS
- after rolling back to v2.48.1 the problem disappears (tested this independently and separately on both servers)
- nothing else was changed when rolling back (database, config file, all the same - I simply stopped the process running v.2.49.x, then started v.2.28.1 and the memory leak problem disappears

I am not able to pinpoint the issue in the code, but there is without a doubt a memory leak problem that was introduced in 2.49

**Comment by mh0lt** [comment]
https://github.com/ledgerwatch/erigon/pull/8305 has a fix for one of the causes of this issue.  I won't close this yet as I'm not sure its the only one.

I think whenever we get delays in processing headers and we're a significant distance from the head of the chain (>1M Blocks) we will see this issue.

Its not clear at the moment if we should fix this to provide a smaller max overall memory footprint (i.e. make the header list size memory dependent) or leave it as is.

**Comment by ryanschneider** [comment]
I can also confirm this issue started happening for us when we upgraded from `23.9.1` to `2.49.3`, now our nodes OOM every 6d or so.  Furthermore we are seeing this issue across all three Ethereum L1 networks (mainnet, Sepolia, goerli) and it looks like #8305 is specific to Polygon so I doubt the issue is really resolved.

A couple of interesting data points:

- One of our nodes was reporting using roughly 70GB of memory for `erigon`, yet Go's `pprof` tool only reported around 5GB in use.

This same node stopped importing new blocks with this error:

```
Oct 21 07:58:13 ip-10-0-65-144 erigon[10575]: [INFO] [10-21|07:58:13.293] head updated                             hash=0x94ed128805345b3244b32a21b019862c13af24e96df41fd06ee1fa4c123ce065 number=18397343
Oct 21 07:58:24 ip-10-0-65-144 erigon[10575]: [INFO] [10-21|07:58:24.109] [NewPayload] Handling new payload        height=18397344 hash=0x2a9873afd92f30005a96f4a904d36278c01464ce9fc5a63df8b4ebb884988153
Oct 21 07:58:32 ip-10-0-65-144 erigon[10575]: [EROR] [10-21|07:58:32.240] RPC method engine_newPayloadV2 crashed: fail to open mdbx: mdbx_env_open: cannot allocate memory, label: inMem, trace: [kv_mdbx.go:263 kv_mdbx.go:396 memory_mutation.go:47 fork_validator.go:150 ethereum_execution.go:181 execution_client.go:59 chain_reader.go:288 engine_server.go:721 engine_server.go:234 engine_server.go:573 value.go:586 value.go:370 service.go:224 handler.go:494 handler.go:444 handler.go:392 handler.go:223 handler.go:316 asm_amd64.s:1598]
Oct 21 07:58:32 ip-10-0-65-144 erigon[10575]: [service.go:219 panic.go:884 kv_mdbx.go:398 memory_mutation.go:47 fork_validator.go:150 ethereum_execution.go:181 execution_client.go:59 chain_reader.go:288 engine_server.go:721 engine_server.go:234 engine_server.go:573 value.go:586 value.go:370 service.go:224 handler.go:494 handler.go:444 handler.go:392 handler.go:223 handler.go:316 asm_amd64.s:1598]
Oct 21 07:58:32 ip-10-0-65-144 erigon[10575]: [WARN] [10-21|07:58:32.240] Served                                   conn=127.0.0.1:52952 method=engine_newPayloadV2 reqid=201088 t=8.133393463s err="method handler crashed"
```

This leads me to believe that the issue is in erigon's use of MBDX, since Go is not seeing the memory from pprof its most likely a leak in the buffers managed by the MBDX C/C++ code.

**Comment by AskAlexSharov** [comment]
@racytech your issue fixed by several prs: 
https://github.com/ledgerwatch/erigon/pull/8434 and https://github.com/ledgerwatch/erigon/pull/8141 and https://github.com/ledgerwatch/erigon/pull/8445 and https://github.com/ledgerwatch/erigon/pull/8490

**Comment by ryanschneider** [comment]
@AskAlexSharov has a new release been tagged with all the above fixes?

**Comment by battlmonstr** [comment]
A few weeks ago I've managed to fix a major memory leak related to failed header verification:
https://github.com/ledgerwatch/erigon/pull/8431
It still leaks 😓 , but at a much slower pace (+10% in 10 minutes compared to +300% in 2 minutes before).

The fix was merged to devel on 2023.10.14 and included in Erigon v2.53.0 release.

Please test again on a new version, and post a new issue if the results are still not satisfactory. 

## Analysis

A header is somewhere between 1400 and 2200 bytes.
The expected Go memory usage with linkLimit=1M at this stage should be around 2.2Gb. 
If the Headers stage prints some progress with memory usage alloc=... and we divide it by linkLimit (1M), and the result is significantly greater than 2200, it means it leaks.

Overall the memory there is structured like so:
* 2-3 Gb - Go runtime resident size (mostly not inserted headers)
* 0-2 Gb - Go runtime preallocated size (it grows when some headers get inserted)
* 3.2 Gb - mdbx "dirty space" controlled by the dirtySpace option: https://github.com/ledgerwatch/erigon/blob/devel/erigon-lib/kv/mdbx/kv_mdbx.go#L88 . Dirty space max size depends on total RAM: 64 Gb total sets 3.2 Gb dirty space, 32 Gb - 1.6 Gb, 16 Gb - 0.8 Gb. In our setup mdbx transaction writes have 3 stages: first the changes are recorded in memory in this "dirty space", then "spilled" to disk if "dirty space" gets full, and then committed (copied?) to the permanent db location. The final commit  happens when StageLoopIteration ends (all stages pass one iteration), it is not happening during headers download stage. So when headers are "inserted", it just moves it from (1) to (3) in a way, and maybe moves some "dirty" stuff to disk.
* 0.8-1 Gb - unknown native memory. This stays relatively constant around 1Gb (doesn't leak).

The expected total required memory is around 8 Gb for the headers stage on a 64 Gb VM, and 6 Gb on a 16 Gb VM.
